text/0000-self-macro-metavar.md

```diff
@@ -0,0 +1,247 @@
+}
+```
+
+On expansion of `m`, `PRIVATE` unambiguously and unconditionally resolves as if
+it were at the definition site, that is, to `submod::PRIVATE`.
+
+[declarative macros 2.0]: https://github.com/rust-lang/rust/issues/39412 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `$self` macro metavariable, like the `$crate` metavariable, can be used to
+modify the hygeine of identifiers in a macro. `$self` works a lot like the
+`self` in module paths: when used at the start of a path in a macro, the
+succeeding path will be resolved as if it were in the module where the macro is
+defined, regardless of where the macro is expanded. Different from `self` in
+module paths, however, `$self` _also_ captures the visibility of the module path
+at the definition site: the succeeding path will be visible in the expansion if
+it is visible at the macro's definition site.
+
+Said differently, `$self` _captures_ the module scope at the macro definition
+site and applies it to the succeeding path upon expansion. As an example,
+consider the definition of the macro `submod::m!`:
+
+```rust
+mod submod {
+    static PRIVATE: &'static str = "PRIVATE_SUBMOD";
+
+    #[macro_export]
+    macro_rules! m {
+        () => (println!("{}", $self::PRIVATE))
+    }
+}
+
+pub fn main() {
+    submod::m!(); // `PRIVATE` unconditionally resolves to `submod::PRIVATE`
+}
+```
+
+Without `$self`, it would not be possible to reference `submod::PRIVATE` outside
+of `submod`. Observe, too, that unlike `$crate`, `$self` _does_ have an effect
+on visibility: while `submod::PRIVATE` in `main` would _not_ resolve, the
+expansion including `$self::PRIVATE` does!
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+At its core, `$self` is `$crate` at the module-level as opposed to the crate
+level. Macro metavariable naming collisions are handled in the same way as with
+`$crate`. In particular, a declaration of `$self` in a macro shadows the `$self`
+described here. The following works as expected, and importantly, as it does
+today:
+
+```rust
+macro_rules! m {
+    ($self:ident) => (println!("{}", $self))
+}
+```
+
+Additionally, like `$crate`, a non-user-declared `$self` _must_ be followed by
+`::`.
+
+Notably different is that while `$crate` can be implemented as a purely
+syntactic transformation, substituting `$crate` for the name of the crate in
+which the macro is defined, `$self` must apply the full resolution context of
+the macro's definition site to the succeeding path. When calling a macro using
+`$self` cross-crate, this requires cross-crate hygiene. Thankfully, this was
+recently added to the compiler in https://github.com/rust-lang/rust/pull/72121.
+
+Thus, `$self` can be simply and without further caveats by specified as: for
+every path in the expansion that begins with `$self`, the resolution context of
+the path is set to resolution context of the `Span::source()` of `$self`.
```

Comment:
Instead of `Span::source()`, I think it would make more sense to talk about `Span::def_site()`.

Reply:
Unfortunately, I don't think `Span::def_site()` is quite right either. For example, in the expansion chain `#[attr] -> macro_rules foo!() { $self }`, `$self` should resolve wherever `#[attr]` was called (or equivalently, where the tokens `macro_rules foo!() { $self }` expanded to), not where `#[attr]` was defined.

I _believe_ the precise way to phrase this is:

> Thus, `$self` can simply and without further caveats by specified as: for
> every path in the expansion that begins with `$self`, the resolution context of
> the path is set to resolution context of the first `macro_rules!` definition in which
> the `$self` tokens appeared.

Then proceed with show, as examples, 1) nested `macro_rules!`, where the _outer_ scope is used, and 2) a `proc-macro` that expands to a `macro_rules!`, where the call-site of the `proc-macro` is used.

Reply:
Sorry for being unclear  - I meant that the `$self` path should be resolved as if it was produced by a proc-macro with `Span::def_site()` (after the parent module path has been substituted).